pkg/controller/cluster/agent/shared.go

```diff
@@ -0,0 +1,249 @@
+package agent
+
+import (
+	"fmt"
+
+	"github.com/rancher/k3k/pkg/apis/k3k.io/v1alpha1"
+	"github.com/rancher/k3k/pkg/controller"
+	apps "k8s.io/api/apps/v1"
+	v1 "k8s.io/api/core/v1"
+	rbacv1 "k8s.io/api/rbac/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrlruntimeclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	sharedKubeletConfigPath = "/opt/rancher/k3k/config.yaml"
+	sharedNodeAgentName     = "kubelet"
+	SharedNodeMode          = "shared"
+)
+
+type SharedAgent struct {
+	cluster          *v1alpha1.Cluster
+	serviceIP        string
+	sharedAgentImage string
+}
+
+func NewSharedAgent(cluster *v1alpha1.Cluster, serviceIP, sharedAgentImage string) Agent {
+	return &SharedAgent{
+		cluster:          cluster,
+		serviceIP:        serviceIP,
+		sharedAgentImage: sharedAgentImage,
+	}
+}
+
+func (s *SharedAgent) Config() (ctrlruntimeclient.Object, error) {
+	config := sharedAgentData(s.cluster)
+
+	return &v1.Secret{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "Secret",
+			APIVersion: "v1",
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      configSecretName(s.cluster.Name),
+			Namespace: s.cluster.Namespace,
+		},
+		Data: map[string][]byte{
+			"config.yaml": []byte(config),
+		},
+	}, nil
+}
+
+func sharedAgentData(cluster *v1alpha1.Cluster) string {
+	nodeName := cluster.Name + "-" + "k3k-kubelet"
+	return fmt.Sprintf(`clusterName: %s
+clusterNamespace: %s
+nodeName: %s
+agentHostname: %s
+token: %s`,
+		cluster.Name, cluster.Namespace, nodeName, nodeName, cluster.Spec.Token)
+}
+
+func (s *SharedAgent) Resources() ([]ctrlruntimeclient.Object, error) {
+	var objs []ctrlruntimeclient.Object
+	objs = append(objs, s.serviceAccount(), s.role(), s.roleBinding(), s.service(), s.deployment())
+	return objs, nil
```

Comment:
```suggestion
	return []ctrlruntimeclient.Object{s.serviceAccount(), s.role(), s.roleBinding(), s.service(), s.deployment()}, nil
```

Not sure if this will work but makes sense and will remove a line from the function.